File: tree/다단계칫솔판매.py
#https://programmers.co.kr/learn/courses/30/lessons/77486?language=python3


def solution(enroll, referral, seller, amount):
    N = len(enroll)
    answer = []
    p = {'root' : 'root'}
    moneys = {'root' : 0}
    
    # 돈, 부모 초기화
    for e, pe in zip(enroll,referral):
        if pe != '-':
            p[e] = pe
        else:
            p[e] = 'root'
        moneys[e] = 0
    
    def go(x, cost):
        if x == 'root':
            return 
        temp = cost // 10
        moneys[x] -= temp
        moneys[p[x]] += temp
        if temp == 0:
            return
        go(p[x], temp)
    
    for s, cnt in zip(seller, amount):
        cost = cnt * 100
        moneys[s] += cost        
        go(s, cost)
        
    for e in enroll:
        answer.append(moneys[e])
        
    return answer

File: tree/다단계칫솔판매.java
import java.util.*;
import java.util.Map.Entry;

class Solution {
    HashMap<String, String> parents = new HashMap<>();
    HashMap<String, Integer> money_map = new HashMap<>();
    
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        int[] answer = new int[enroll.length];
        
        for(int i=0; i< enroll.length; i++){
            String en = enroll[i];
            parents.put(en, referral[i]);
            money_map.put(en, 0);
        }
        
        for(int i=0; i< seller.length; i++){
            int temp = amount[i] * 100;
            go(seller[i], temp);
        }
        
        for(int i=0; i< enroll.length; i++){
            String en = enroll[i];
            answer[i] = money_map.get(en);
        }
        
        return answer;
    }
    
    
    
    public void go(String user, int money){
        String parent = parents.get(user);
        if(user.equals("-")){
            return;
        }
        
        int temp = money / 10;
        money_map.put(user, money_map.get(user)+money-temp);
        if(temp == 0){
            return;
        }
        go(parent, temp);
    }
    
}